internal/controllers/httpsedge_controller.go

```diff
@@ -189,71 +197,91 @@ func (r *HTTPSEdgeReconciler) reconcileRoutes(ctx context.Context, edge *ingress
 					r.Recorder.Eventf(edge, v1.EventTypeWarning, "FailedValidate", "Could not validate ip restriction: %v", err)
 					continue
 				}
-
-				return err
+				return errors.NewErrInvalidConfiguration(err.Error())
 			}
 		}
 
 		match := r.getMatchingRouteFromEdgeStatus(edge, routeSpec)
 		var route *ngrok.HTTPSEdgeRoute
+		// Now we go ahead and create the route if it doesn't exist.
+		// It's important to note here that we are intentionally ommiting the `route.Backend` for new routes.
+		//  The success or failure of applying a route's modules is then strongly linked the state of its backend.
+		//  Thus, any route with a backend is considered properly configured.
+		//  See https://github.com/ngrok/kubernetes-ingress-controller/issues/208 for additional context.
 		if match == nil {
-			r.Log.Info("Creating new route", "edgeID", edge.Status.ID, "match", routeSpec.Match, "matchType", routeSpec.MatchType, "backendID", backend.ID)
-			// This is a new route, so we need to create it
+			r.Log.Info("Creating new route", "edgeID", edge.Status.ID, "match", routeSpec.Match, "matchType", routeSpec.MatchType)
 			req := &ngrok.HTTPSEdgeRouteCreate{
 				EdgeID:    edge.Status.ID,
 				Match:     routeSpec.Match,
 				MatchType: routeSpec.MatchType,
-				Backend: &ngrok.EndpointBackendMutate{
-					BackendID: backend.ID,
-				},
 			}
-			route, err = r.NgrokClientset.HTTPSEdgeRoutes().Create(ctx, req)
+			route, err = edgeRoutes.Create(ctx, req)
+
+			// Update status for newly created route
+			routeStatuses[i] = ingressv1alpha1.HTTPSEdgeRouteStatus{
+				ID:        route.ID,
+				URI:       route.URI,
+				Match:     route.Match,
+				MatchType: route.MatchType,
+			}
 		} else {
-			r.Log.Info("Updating route", "edgeID", edge.Status.ID, "match", routeSpec.Match, "matchType", routeSpec.MatchType, "backendID", backend.ID)
-			// This is an existing route, so we need to update it
-			req := &ngrok.HTTPSEdgeRouteUpdate{
-				EdgeID:    edge.Status.ID,
-				ID:        match.ID,
-				Match:     routeSpec.Match,
-				MatchType: routeSpec.MatchType,
-				Backend: &ngrok.EndpointBackendMutate{
-					BackendID: backend.ID,
-				},
+			req := &ngrok.EdgeRouteItem{
+				ID:     match.ID,
+				EdgeID: edge.Status.ID,
 			}
-			route, err = r.NgrokClientset.HTTPSEdgeRoutes().Update(ctx, req)
+			route, err = edgeRoutes.Get(ctx, req)
 		}
 		if err != nil {
 			return err
 		}
-		routeStatuses[i] = ingressv1alpha1.HTTPSEdgeRouteStatus{
+
+		// With the route properly staged, we now attempt to apply its module updates
+		// TODO: Check if there are no updates to apply here to skip any unnecessary disruption
+		r.Log.Info("Applying route modules", "edgeID", edge.Status.ID, "match", routeSpec.Match, "matchType", routeSpec.MatchType)
+		if err := routeModuleUpdater.updateModulesForRoute(ctx, route, &routeSpec); err != nil {
+			r.Recorder.Event(edge, v1.EventTypeWarning, "RouteModuleUpdateFailed", err.Error())
+			return errors.NewErrInvalidConfiguration(err.Error())
```

Comment:
This seems like it turns any ngrok api error into an unretriable "invalid configuration" error.

If I understand correctly, that means a network timeout, or ngrok 5xx, or whatever, would result in us giving up on the update and leaving things in a broken state.

Shouldn't we be retrying network issues, and looking for specific ngrok error codes that indicate invalid configuration?

Reply:
Hmm, fair point. I suppose each individual module call would have to be config error aware.

Reply:
It seems like there's a _whole lot_ of [possibilities](https://github.com/ngrok-private/ngrok/blob/0827b5cf976a368dc384b673e560948d900fbfef/js/packages/code-gen/error-code/index.ts) that we could get back, and the errors don't seem to contain that much information (just the error message itself).

Is there a sane way to match against what we could consider a "configuration" error?